lib/Driver/UnixToolChains.cpp

```diff
@@ -108,8 +110,22 @@ ToolChain::InvocationInfo toolchains::GenericUnix::constructInvocation(
 
   return II;
 }
-
+// Amazon Linux 2023 requires lld as the default linker.
 std::string toolchains::GenericUnix::getDefaultLinker() const {
+  if (getTriple().isOSLinux()) {
+    std::ifstream file("/etc/os-release");
```

Comment:
Got a better way to organize this: can you [setup an AmazonLinux toolchain driver instead, like OpenBSD did](https://github.com/apple/swift/blob/ef8825b/lib/Driver/ToolChains.h#L198), [instantiate that based on the vendor](https://github.com/apple/swift/blob/ef8825b/lib/Driver/Driver.cpp#L356), then [add whatever logic like this that you need in a `getDefaultLinker()` override instead](https://github.com/apple/swift/blob/bf137cb30dda404ffc6b15658e61a4f41187d2cb/lib/Driver/UnixToolChains.cpp#L460)? That would be cleaner.

Reply:
I will take a look at your suggestion, but as the move is to make `lld` the default linker hopefully none of this code will be needed soon.

Reply:
Sure, if the Swift devs say they will be removing `lib/Driver/` altogether and don't need it changed too, just the CMake change should be enough.

Reply:
I have had a look at this and I am not sure how this could work in the case of Amazon Linux.
With `OpenBSB` we are dealing with a specific `OSType` where as `Amazon Linux` is just one of the many `Linux` distributions. 
There is no `VendorType`, `OSType` or `EnvironmentType` that can differentiate between different Linux distributions.
As we don't want to effect any other of the Linux distributions I think the code I have put forward is the simplest way to achieve this.

Reply:
> There is no VendorType, OSType or EnvironmentType that can differentiate between different Linux distributions.

You are right: I assumed there would be, like [there is one for `SUSE`, but there isn't one for Amazon](https://github.com/apple/llvm-project/blob/stable/20230725/llvm/include/llvm/TargetParser/Triple.h#L166).

> As we don't want to effect any other of the Linux distributions I think the code I have put forward is the simplest way to achieve this.

I don't think this is interfering as little as it should though. I suggest you spin this logic off as a separate internal method `bool isAmazonLinux2023Host()`, then simply modify the Android line below: `if (getTriple().isAndroid() || isAmazonLinux2023Host())`. That way, at least nobody has to read all this logic if they don't care about AL2023.

Reply:
That sounds like a good idea, I will make those changes.